unison-share-api/src/Unison/Server/CodebaseServer.hs

```diff
@@ -146,16 +147,72 @@ data BaseUrl = BaseUrl
     urlPort :: Port
   }
 
-data BaseUrlPath = UI | Api
+data NameOrHashRef = Name [Text] | Hash Text
+
+data DefinitionReference
+  = TermReference NameOrHashRef
+  | TypeReference NameOrHashRef
+  | AbilityConstructorReference NameOrHashRef
+  | DataConstructorReference NameOrHashRef
+
+data BaseUrlPath
+  = UI
+      { namespace :: Maybe [Text],
+        definition :: Maybe DefinitionReference
+      }
+  | Api
 
 instance Show BaseUrl where
   show url = urlHost url <> ":" <> show (urlPort url) <> "/" <> (URI.encode . unpack . urlToken $ url)
 
 urlFor :: BaseUrlPath -> BaseUrl -> String
-urlFor path baseUrl =
-  case path of
-    UI -> show baseUrl <> "/ui"
+urlFor baseUrlPath baseUrl =
+  case baseUrlPath of
+    UI ns def ->
+      show baseUrl <> "/ui" <> Text.unpack (path ns def)
     Api -> show baseUrl <> "/api"
+  where
+    path :: Maybe [Text] -> Maybe DefinitionReference -> Text
+    path ns def =
+      case (namespacePath ns, definitionPath def) of
+        (Just nsPath, Just defPath) -> "/latest" <> nsPath <> ";" <> defPath
+        (Just nsPath, Nothing) -> "/latest" <> nsPath
+        (Nothing, Just defPath) -> "/latest" <> defPath
+        (Nothing, Nothing) -> ""
+
+    namespacePath :: Maybe [Text] -> Maybe Text
+    namespacePath ns =
+      fmap (("/namespaces/" <>) . toPath) ns
+
+    definitionPath :: Maybe DefinitionReference -> Maybe Text
+    definitionPath def =
+      toDefinitionPath <$> def
+
+    toPath :: [Text] -> Text
+    toPath parts =
+      parts
+        & fmap UriEncode.encodeText
+        & Text.intercalate "/"
```

Comment:
I'm biased because of `|>` — let me know if I stop this nonsense :D

Reply:
We use `&` all over :+1: